src/UndockedRegFreeWinRT/UndockedRegFreeWinRT/catalog.cpp

```diff
@@ -81,72 +82,373 @@ struct component
 };
 
 static unordered_map<wstring, shared_ptr<component>> g_types;
+static unordered_set<wstring> g_assemblies;
 
-HRESULT WinRTLoadComponent(PCWSTR manifest_path)
+HRESULT LoadFromSxSManifest(std::wstring path)
+{
+    return WinRTLoadComponentFromFilePath(path.c_str());
+}
+
+HRESULT LoadFromEmbeddedManifest(std::wstring path)
+{
+    int resource = 0;
+    if (path.size() < 4)
+    {
+        return COR_E_ARGUMENT;
+    }
+    std::wstring ext = path.substr(path.size() - 4, path.size());
+    if (ext.compare(L".exe") == 0)
+    {
+        resource = 1;
+    }
+    else if (ext.compare(L".dll") == 0)
+    {
+        // I notice in testing, the original sxs allows this number to be also 1. 
+        resource = 2;
+    }
+    else
+    {
+        return COR_E_ARGUMENT;
+    }
+
+    HMODULE handle = LoadLibraryExW(path.c_str(), nullptr, LOAD_LIBRARY_AS_DATAFILE_EXCLUSIVE);
+    if (!handle)
+    {
+        return ERROR_FILE_NOT_FOUND;
+    }
+
+    HRSRC hrsc = FindResourceW(handle, MAKEINTRESOURCEW(resource), RT_MANIFEST);
+    if (!hrsc)
+    {
+        return ERROR_FILE_NOT_FOUND;
+    }
+    HGLOBAL embeddedManifest = LoadResource(handle, hrsc);
+    if (!embeddedManifest)
+    {
+        return ERROR_FILE_NOT_FOUND;
+    }
+    DWORD length = SizeofResource(handle, hrsc);
+    void* data = LockResource(embeddedManifest);
+    if (!data)
+    {
+        return ERROR_FILE_NOT_FOUND;
+    }
+    std::string result = std::string((char*)data, length);
+    return WinRTLoadComponentFromString(result.c_str());
+}
+
+HRESULT WinRTLoadComponentFromFilePath(PCWSTR manifestPath)
 {
     ComPtr<IStream> fileStream;
-    ComPtr<IXmlReader> xmlReader;
+    RETURN_IF_FAILED(SHCreateStreamOnFileEx(manifestPath, STGM_READ, FILE_ATTRIBUTE_NORMAL, false, nullptr, &fileStream));
+    return ParseRootManifestFromXmlReaderInput(fileStream.Get());;
+}
+
+HRESULT WinRTLoadComponentFromString(PCSTR xmlStringValue)
+{
+    ComPtr<IStream> xmlStream = nullptr;
+    xmlStream.Attach(SHCreateMemStream(reinterpret_cast<const BYTE*>(xmlStringValue), strlen(xmlStringValue) * sizeof(CHAR)));
```

Comment:
Does this make another copy of the incoming data? I can't quite parse what MSDN's saying.

Reply:
You're talking about SHCreateMemStream right? Yea the MSDN isn't so clear about it.